src/compiler/scala/tools/nsc/typechecker/EtaExpansion.scala

```diff
@@ -55,16 +55,22 @@ trait EtaExpansion { self: Analyzer =>
         else {
           val vname: Name = freshName()
           // Problem with ticket #2351 here
+          val valSym = typer.context.owner.newValue(vname.toTermName, tree.pos.focus, SYNTHETIC)
           defs += atPos(tree.pos) {
             val rhs = if (byName) {
-              val res = typer.typed(Function(List(), tree))
-              new ChangeOwnerTraverser(typer.context.owner, res.symbol) traverse tree // scala/bug#6274
-              res
-            } else tree
-            ValDef(Modifiers(SYNTHETIC), vname.toTermName, TypeTree(), rhs)
+              val funSym = valSym.newAnonymousFunctionValue(tree.pos.focus)
+              val tree1 = tree.changeOwner(typer.context.owner -> funSym)
+              val funType = definitions.functionType(Nil, tree1.tpe)
+              funSym.setInfo(funType)
+              Function(List(), tree1).setSymbol(funSym).setType(funType)
+            } else {
+              tree.changeOwner(typer.context.owner -> valSym)
+            }
+            valSym.setInfo(rhs.tpe)
+            ValDef(valSym, rhs)
```

Comment:
looks good; `def etaExpand` could now take a `owner: Symbol` parameter instead of the full `typer: Typer`.

Reply:
Pushed a commit with that refactoring.